homeassistant/components/nest/__init__.py

```diff
@@ -194,63 +194,71 @@ def __init__(self, hass, conf, nest):
 
     def initialize(self):
         """Initialize Nest."""
-        if self.local_structure is None:
-            self.local_structure = [s.name for s in self.nest.structures]
+        from nest.nest import AuthorizationError, APIError
+        try:
+            # Do not optimize next statement, it is here for initialize
+            # persistence Nest API connection.
+            structure_names = [s.name for s in self.nest.structures]
+            if self.local_structure is None:
+                self.local_structure = structure_names
+
+        except (AuthorizationError, APIError, socket.error) as err:
```

Comment:
Right now we will fail silently. Instead, we should return false from the setup config entry method to indicate setup has failed

Reply:
Changed to re-thrown exception here. Changed setup._async_setup_component to check entry.async_setup result.

But I cannot find test for _async_setup_component, do we have?

Reply:
No. There are no tests for Nest. I've only added a few now to make sure the config flow is covered.

Reply:
Just have initialize return False and then make sure setup_config_entry returns False and stops execution.

Reply:
Are you sure? It still fail silently even I return False, nobody check return of entry.async_setup, like I changed in setup.py.

This initialize function is not part of config flow at all, it will be called after config flow finished.

To test it, I modified the access token in `.config_entries.json`, if I only return False, I will get logs likes
```
2018-06-16 21:59:32 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=camera>
2018-06-16 21:59:32 ERROR (SyncWorker_1) [homeassistant.components.nest] Connection error while access Nest web service: Authorization Failed
2018-06-16 21:59:33 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=nest>
2018-06-16 21:59:33 INFO (MainThread) [homeassistant.bootstrap] Home Assistant initialized in 1.65s
2018-06-16 21:59:33 INFO (MainThread) [homeassistant.core] Starting Home Assistant core loop
2018-06-16 21:59:33 INFO (MainThread) [homeassistant.core] Starting Home Assistant
```

If I applied setup.py changes, I will get
```
2018-06-16 22:02:24 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=config>
2018-06-16 22:02:26 ERROR (SyncWorker_1) [homeassistant.components.nest] Connection error while access Nest web service: Authorization failed
2018-06-16 22:02:26 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=nest>
2018-06-16 22:02:26 INFO (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=persistent_notification, service=create, service_data=title=Invalid config, message=The following components and platforms could not be set up:

 - [nest](https://home-assistant.io/components/nest/)

Please check your config., notification_id=invalid_config, service_call_id=139973454937952-1>
```
Then I will get notification card in frontend, will it better then just an error log?
![image](https://user-images.githubusercontent.com/7366491/41504940-230adff8-71b2-11e8-9439-364a7b1d42d3.png)

Reply:
Where is initialize called? A couple of lines above inside the `async_setup_entry` method on line 125. Just check the return value there and return False if it was unable to initialize.

Reply:
Yes, that is what I have done in previous comment. Nobody check aysnc_setup_entry's return value, so there is still quietly failed to user.

Reply:
The status is shown on the integrations page in the config. We don't have a way yet to specify the error message.

Reply:
Changed to return False in except block